### PR TITLE
PN-401: changes to the match notification DB layer

### PR DESCRIPTION
**Pull request by allasm**
Initial implementation, needs PN-453 and PN-456 to be complete. PR is made for code review purposes only and to simplify testing this part of the code

Should be tested together with RM-129

@sdumitriu: when upgrading existing instances (e.g. prod) need to drop all existing patient_matching table indexes first

### CODE REVIEW

File: matching-notification-api/src/main/java/org/phenotips/matchingnotification/match/PatientMatch.java [hunk 1]
```diff
@@ -351,7 +339,7 @@
     /**
      * Add a record to the notification history in case user contacted regarding this match outside of PhenomeCentral.
      *
-     * @param isUserContacted boolean indicator whether contact regarding this match was done or not
+     * @param isExternallyContacted boolean indicator whether contact regarding this match was done or not
```
Comment (veronikaslc) [inline matching-notification-api/src/main/java/org/phenotips/matchingnotification/match/PatientMatch.java]: unrelated change to the issue

File: matching-notification-api/src/main/java/org/phenotips/matchingnotification/match/internal/DefaultPatientMatch.java [hunk 2]
```diff
@@ -714,6 +693,7 @@ public JSONObject getNotificationHistory()
     public void setNotificationHistory(JSONObject notificationHistory)
     {
         this.notificationHistory = (notificationHistory != null) ? notificationHistory.toString() : null;
+        this.notifiedTimestamp = new Timestamp(System.currentTimeMillis());
```
Comment (veronikaslc) [inline matching-notification-api/src/main/java/org/phenotips/matchingnotification/match/internal/DefaultPatientMatch.java]: wring change, `notifiedTimestamp` with be set to current time each time we call `setNotificationHistory()` in `preserveOriginalMatchMetaInfo()`

File: similarity-data-rest/src/main/java/org/phenotips/data/similarity/rest/internal/DefaultPatientMatchResource.java [hunk 3]
```diff
@@ -131,11 +134,17 @@ private Response generateMatchesResponse(
     {
         final List<PatientSimilarityView> matches = this.similarPatientsFinder.findSimilarPatients(patient);
 
-        this.matchStorageManager.saveLocalMatches(matches, patient.getId());
+        Map<PatientSimilarityView, PatientMatch> savedMatches =
+            this.matchStorageManager.saveLocalMatches(matches, patient.getId());
 
-        final MatchedPatientClusterView cluster = new DefaultMatchedPatientClusterView(patient, matches);
+        @SuppressWarnings({ "rawtypes", "unchecked" })
```
Comment (allasm) [inline similarity-data-rest/src/main/java/org/phenotips/data/similarity/rest/internal/DefaultPatientMatchResource.java]: I think you don't need to suppress any warnings if you do `new HashMap<>()`; on the next line

File: matching-notification-api/src/main/java/org/phenotips/matchingnotification/match/internal/DefaultPatientMatch.java (hunk 2 above)
Comment (allasm) [reply matching-notification-api/src/main/java/org/phenotips/matchingnotification/match/internal/DefaultPatientMatch.java]: I think that is OK. It is not really used anywhere (we have notification history if we need details), and in the next update we should probably remove it completely

File: matching-notification-api/src/main/java/org/phenotips/matchingnotification/storage/internal/DefaultMatchStorageManager.java [hunk 4]
```diff
@@ -138,125 +103,166 @@
...
-                if (!patientId.equals(refPatientID)) {
-                    refPatients.add(refPatientID);
-                    if (isIncoming) {
-                        this.logger.error(
-                            "A list of incoming matches for remote patient {} also constains matches for patient {}",
-                            patientId, refPatientID);
+
+            for (Map.Entry<PatientSimilarityView, PatientMatch> entry : matchMapping.entrySet()) {
+                PatientMatch match = entry.getValue();
+                if (matchesByMatchedPatient.containsKey(match.getMatchedPatientId())) {
```
Comment (veronikaslc) [inline matching-notification-api/src/main/java/org/phenotips/matchingnotification/storage/internal/DefaultMatchStorageManager.java]: NPE when `match = null` from [this](https://github.com/phenotips/patient-network/pull/261/files#diff-01f699dad56819dad6cc877f31af51cbR140) line

File: matching-notification-api/src/main/java/org/phenotips/matchingnotification/storage/internal/DefaultMatchStorageManager.java [hunk 5]
```diff
@@ -138,125 +103,166 @@
...
-        for (PatientSimilarityView similarityView : inputMatches) {
-            PatientMatch match = new DefaultPatientMatch(similarityView, null, null);
-
-            // filter out matches owned by the same user(s), as those are not shown in matching notification anyway
-            // and they break match count calculation if they are included
-            if (match.getReference().getEmails().size() > 0 && CollectionUtils.isEqualCollection(
-                match.getReference().getEmails(), match.getMatched().getEmails())) {
-                continue;
-            }
+        Predicate<PatientMatch> filterOutSameOwnerMatches = match -> {
```
Comment (veronikaslc) [inline matching-notification-api/src/main/java/org/phenotips/matchingnotification/storage/internal/DefaultMatchStorageManager.java]: why do we need this again after [this](https://github.com/phenotips/patient-network/blob/master/similarity-search/src/main/java/org/phenotips/similarity/internal/SolrSimilarPatientsFinder.java#L216-L220)?

Comment (allasm) [reply matching-notification-api/src/main/java/org/phenotips/matchingnotification/storage/internal/DefaultMatchStorageManager.java]: Ah, this is a regression that we have it "there". The filter is only supposed to be here. I will fix that

File: matching-notification-migrations/src/main/java/org/phenotips/matchingnotification/internal/R71502PatientNetwork196DataMigration.java [hunk 6]
```diff
@@ -134,7 +134,7 @@ public void hibernateMigrate() throws DataMigrationException, XWikiException
                     // re-compute the match with just fetched two local patients
                     PatientSimilarityView similarityView = this.factory.makeSimilarPatient(matchedPatient.getPatient(),
                         referencePatient.getPatient());
-                    PatientMatch newMatch = new DefaultPatientMatch(similarityView, null, null);
+                    PatientMatch newMatch = new CurrentPatientMatch(similarityView, null, null);
```
Comment (veronikaslc) [inline matching-notification-migrations/src/main/java/org/phenotips/matchingnotification/internal/R71502PatientNetwork196DataMigration.java]: @marta- @sdumitriu PN-401 changes `PatientMatch` class implementation and breaks `PatientMatch.class.getCanonicalName()` in other migrators. They also have to be changed to use `CurrentPatientMatch` instead, even if we think we will never run them again. Not a big change, but we will have not broken code in repo.

**Comment by veronikaslc** [comment]
`[misc]` commit is unrelated, actually changes from that commit are a part of another PR for PN-483, so commit should be removed from this PR

File: matching-notification-migrations/src/main/java/org/phenotips/matchingnotification/internal/R74696PatientNetwork456DataMigration.java [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+    // the "d.comments IS NULL" check is a safeguard, since in the production DB there should be no matches
+    // that have comments and have a notified copy of a match (and in fact there are none as of last check)
+    private static final String SQL_DELETE_NON_NOTIFIED_COPIES_OF_A_NOTIFIED_MATCH =
+        "delete from patient_matching d where d.comments IS NULL and d.notified = false and exists"
+        + " (select * from patient_matching d2 where"
+        + " d2.referencePatientId = d.referencePatientId and d2.matchedPatientId = d.matchedPatientId"
+        + " and d2.matchedServerId = d.matchedServerId and d2.referenceServerId = d.referenceServerId"
+        + " and d2.notified = true)";
+
+    private static final String HQL_DELETE_DUPLICATE_LOCAL_MATCHES =
```
Comment (veronikaslc) [inline matching-notification-migrations/src/main/java/org/phenotips/matchingnotification/internal/R74696PatientNetwork456DataMigration.java]: why we don't really need HQL here since we already have to use SQL

File: matching-notification-migrations/src/main/java/org/phenotips/matchingnotification/internal/R74696PatientNetwork456DataMigration.java [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+ */
+@Component
+@Named("R74696PatientNetwork456")
+@Singleton
+public class R74696PatientNetwork456DataMigration extends AbstractHibernateDataMigration
+{
+    // the "d.comments IS NULL" check is a safeguard, since in the production DB there should be no matches
+    // that have comments and have a notified copy of a match (and in fact there are none as of last check)
+    private static final String SQL_DELETE_NON_NOTIFIED_COPIES_OF_A_NOTIFIED_MATCH =
+        "delete from patient_matching d where d.comments IS NULL and d.notified = false and exists"
```
Comment (veronikaslc) [inline matching-notification-migrations/src/main/java/org/phenotips/matchingnotification/internal/R74696PatientNetwork456DataMigration.java]: ... and what do we do with matches if `d.comments ` is not `NULL`?

File: matching-notification-migrations/src/main/java/org/phenotips/matchingnotification/internal/R74696PatientNetwork456DataMigration.java [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+
+    private static final String HQL_DELETE_DUPLICATE_LOCAL_MATCHES =
+        "delete from CurrentPatientMatch d where d.matchedServerId = '' and d.referenceServerId = ''"
+        + " and d.referencePatientId > d.matchedPatientId and exists"
+        + " (from CurrentPatientMatch d2 where"
+        + " d2.referencePatientId = d.matchedPatientId and d2.matchedPatientId = d.referencePatientId"
+        + " and d2.matchedServerId = '' and d2.referenceServerId = '')";
+
+    private static final String NORMALIZE_ORDER_OF_PATIENTS_IN_A_MATCH =
+        "update CurrentPatientMatch set referencePatientId = matchedPatientId, matchedPatientId = referencePatientId"
```
Comment (veronikaslc) [inline matching-notification-migrations/src/main/java/org/phenotips/matchingnotification/internal/R74696PatientNetwork456DataMigration.java]: runs on all patient matches, has to either run on local matches only, or switch server IDs.

File: matching-notification-migrations/src/main/java/org/phenotips/matchingnotification/internal/R74696PatientNetwork456DataMigration.java [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+import org.slf4j.Logger;
+
+import com.xpn.xwiki.XWikiException;
+import com.xpn.xwiki.store.hibernate.HibernateSessionFactory;
+import com.xpn.xwiki.store.migration.DataMigrationException;
+import com.xpn.xwiki.store.migration.XWikiDBVersion;
+import com.xpn.xwiki.store.migration.hibernate.AbstractHibernateDataMigration;
+
+/**
+ * Migration for PatientNetwork issue PN-456: local matches should be de-duplicated:
```
Comment (veronikaslc) [inline matching-notification-migrations/src/main/java/org/phenotips/matchingnotification/internal/R74696PatientNetwork456DataMigration.java]: Description should be updated to include changes from `SQL_DELETE_NON_NOTIFIED_COPIES_OF_A_NOTIFIED_MATCH`

File: matching-notification-migrations/src/main/java/org/phenotips/matchingnotification/internal/R71502PatientNetwork196DataMigration.java (hunk 6 above)
Comment (allasm) [reply matching-notification-migrations/src/main/java/org/phenotips/matchingnotification/internal/R71502PatientNetwork196DataMigration.java]: Ads discussed, we will ignore older migrators for PC

**Comment by allasm** [comment]
Rebased on latest master

File: matching-notification-migrations/src/main/java/org/phenotips/matchingnotification/internal/R74696PatientNetwork456DataMigration.java [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+import org.slf4j.Logger;
+
+import com.xpn.xwiki.XWikiException;
+import com.xpn.xwiki.store.hibernate.HibernateSessionFactory;
+import com.xpn.xwiki.store.migration.DataMigrationException;
+import com.xpn.xwiki.store.migration.XWikiDBVersion;
+import com.xpn.xwiki.store.migration.hibernate.AbstractHibernateDataMigration;
+
+/**
+ * Migration for PatientNetwork issue PN-456: the clean up of the matching notificstion database
```
Comment (sdumitriu) [inline matching-notification-migrations/src/main/java/org/phenotips/matchingnotification/internal/R74696PatientNetwork456DataMigration.java]:
```suggestion
 * Migration for PatientNetwork issue PN-456: the clean up of the matching notification database
```

File: matching-notification-migrations/src/main/java/org/phenotips/matchingnotification/internal/R74696PatientNetwork456DataMigration.java [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+    @Override
+    public String getDescription()
+    {
+        return "Remove duplicate matches and normalize order of patients in a local match";
+    }
+
+    @Override
+    public XWikiDBVersion getVersion()
+    {
+        return new XWikiDBVersion(74697);
```
Comment (sdumitriu) [inline matching-notification-migrations/src/main/java/org/phenotips/matchingnotification/internal/R74696PatientNetwork456DataMigration.java]: This does not match the version in the classname.

Comment (allasm) [reply matching-notification-migrations/src/main/java/org/phenotips/matchingnotification/internal/R74696PatientNetwork456DataMigration.java]: Right, committed a debug DB version by accident

**Comment by veronikaslc** [comment]
approve this PR
